src/fastpath/fastpath.c

```diff
@@ -472,19 +762,82 @@ void NORETURN fastpath_reply_recv(word_t cptr, word_t msgInfo)
     callerSlot->cteMDBNode = nullMDBNode;
 #endif
 
-    /* I know there's no fault, so straight to the transfer. */
+#ifdef CONFIG_EXCEPTION_FASTPATH
 
+    if (unlikely(fault_type != seL4_Fault_NullFault)) {
+        bool_t restart = 0;
+        switch (fault_type) {
+        case (seL4_Fault_CapFault): {
+            break;
+        }
+        case (seL4_Fault_UnknownSyscall): {
+            break;
+        }
+        case (seL4_Fault_UserException): {
+            break;
+        }
+#ifdef CONFIG_KERNEL_MCS
+        case (seL4_Fault_Timeout): {
+            break;
+        }
+#endif
+
+#ifdef CONFIG_HARDWARE_DEBUG_API
+        case (seL4_Fault_DebugException): {
+            break;
+        }
+#endif
+        default: {
+            restart = 1;
+        }
+        }
+
+        /* TODO: This works here right now because vm faults always restart - the alternative is that the thread becomes inactive,
+         * which would either be handled in the fastpath or redirected into the slowpath (preferred) - either way, this check
+         * will need to be moved above the point of no return to be handled to ensure we don't try and force a switch to an
+         * inactive thread. */
+
+        if (restart) {
+            word_t pc = getRestartPC(caller);
+            setNextPC(caller, pc);
+        }
+
+        /* Clear the tcbFault variable to indicate that it has been handled. */
+        caller->tcbFault = seL4_Fault_NullFault_new();
+
+        /* Dest thread is set Running, but not queued. */
+        thread_state_ptr_set_tsType_np(&caller->tcbState, ThreadState_Running);
+        switchToThread_fp(caller, cap_pd, stored_hw_asid);
+
+        /* The badge/msginfo do not need to be not sent - this is not necessary for exceptions */
+        restore_user_context();
+    } else {
+        /* Replies don't have a badge. */
+        badge = 0;
+
+        fastpath_copy_mrs(length, NODE_STATE(ksCurThread), caller);
+
+        /* Dest thread is set Running, but not queued. */
+        thread_state_ptr_set_tsType_np(&caller->tcbState, ThreadState_Running);
+        switchToThread_fp(caller, cap_pd, stored_hw_asid);
+
+        msgInfo = wordFromMessageInfo(seL4_MessageInfo_set_capsUnwrapped(info, 0));
+
+        fastpath_restore(badge, msgInfo, NODE_STATE(ksCurThread));
+    }
+#else
     /* Replies don't have a badge. */
     badge = 0;
 
     fastpath_copy_mrs(length, NODE_STATE(ksCurThread), caller);
 
     /* Dest thread is set Running, but not queued. */
-    thread_state_ptr_set_tsType_np(&caller->tcbState,
-                                   ThreadState_Running);
+    thread_state_ptr_set_tsType_np(&caller->tcbState, ThreadState_Running);
```

Comment:
please revert style changes, there should be zero code diff when the new fastpath is not configured.